flink-connectors/flink-connector-files/src/test/java/org/apache/flink/connector/file/src/impl/DynamicFileSplitEnumeratorTest.java

```diff
@@ -0,0 +1,189 @@
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.types.logical.RowType;
+
+import org.junit.jupiter.api.Test;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/** Tests for the {@link DynamicFileSplitEnumerator}. */
+class DynamicFileSplitEnumeratorTest {
+
+    @Test
+    void testEnumerating() {
+        String[] splits = new String[] {"0", "1", "2", "3", "4"};
+        MockSplitEnumeratorContext<TestSplit> context = new MockSplitEnumeratorContext<>(1);
+        context.registerReader(new ReaderInfo(0, ""));
+
+        DynamicFileSplitEnumerator<TestSplit> enumerator =
+                new DynamicFileSplitEnumerator<>(
+                        context,
+                        () -> new TestDynamicFileEnumerator(splits, splits),
+                        SimpleSplitAssigner::new);
+
+        for (String ignored : splits) {
+            enumerator.handleSplitRequest(0, null);
+        }
+
+        assertThat(getAssignedSplits(context)).containsExactlyInAnyOrder(splits);
+    }
+
+    @Test
+    void testDynamicFiltering() {
+        String[] splits = new String[] {"0", "1", "2", "3", "4"};
+        String[] remainingSplits = new String[] {"1", "3"};
+        MockSplitEnumeratorContext<TestSplit> context = new MockSplitEnumeratorContext<>(1);
+        context.registerReader(new ReaderInfo(0, ""));
+
+        DynamicFileSplitEnumerator<TestSplit> enumerator =
+                new DynamicFileSplitEnumerator<>(
+                        context,
+                        () -> new TestDynamicFileEnumerator(splits, remainingSplits),
+                        SimpleSplitAssigner::new);
+        enumerator.handleSourceEvent(0, mockDynamicFilteringEvent());
+
+        enumerator.handleSplitRequest(0, null);
+        enumerator.handleSplitRequest(0, null);
+
+        assertThat(getAssignedSplits(context)).containsExactlyInAnyOrder(remainingSplits);
+    }
+
+    @Test
+    void testReceiveDynamicFilteringDataAfterStarted() {
+        String[] splits = new String[] {"0", "1", "2", "3", "4"};
+        String[] remainingSplits = new String[] {"1", "3"};
+        MockSplitEnumeratorContext<TestSplit> context = new MockSplitEnumeratorContext<>(1);
+        context.registerReader(new ReaderInfo(0, ""));
+
+        DynamicFileSplitEnumerator<TestSplit> enumerator =
+                new DynamicFileSplitEnumerator<>(
+                        context,
+                        () -> new TestDynamicFileEnumerator(splits, remainingSplits),
+                        SimpleSplitAssigner::new);
+
+        enumerator.handleSplitRequest(0, null);
+
+        List<String> alreadyAssigned = getAssignedSplits(context);
+
+        enumerator.handleSourceEvent(0, mockDynamicFilteringEvent());
+
+        for (int i = 0; i < 4; i++) {
```

Comment:
it's better the end number is greater than 4, if that we can verity the a split will be assigned twice.